components/site/src/link_checking.rs

```diff
@@ -104,27 +117,32 @@ fn should_skip_by_prefix(link: &str, skip_prefixes: &[String]) -> bool {
     skip_prefixes.iter().any(|prefix| link.starts_with(prefix))
 }
 
-fn get_link_domain(link: &str) -> Result<String> {
+fn get_link_domain(link: &str) -> Result<String, String> {
     return match Url::parse(link) {
         Ok(url) => match url.host_str().map(String::from) {
             Some(domain_str) => Ok(domain_str),
-            None => bail!("could not parse domain `{}` from link", link),
+            None => Err(format!("could not parse domain `{}` from link", link)),
         },
-        Err(err) => bail!("could not parse domain `{}` from link: `{}`", link, err),
+        Err(err) => Err(format!("could not parse domain `{}` from link: `{}`", link, err)),
```

Comment:
Why that change? It should be equivalent to `bail!`

Reply:
I'm happy to reverse that change though, I can see how bailing on the first error is more consistent and leads to less complexity.

Reply:
but I believe the code of `get_link_domain` before and the one with your changes is exactly the same no? You're just returning the error either way in `get_link_domain`

It's fine to report all of them though

Reply:
Whoops, for some reason I was under the impression that bail caused an immediate panic.  Whoops, I'll change it back.

Reply:
Done 05dde1e2

Reply:
> It's fine to report all of them though

Question for you.  Since anyhow always prints an error message, and I'm trying to print multiple error messages at a time, how do you feel about these choices:

1. (Current Zola behavior) Bail out when the very first error is encountered (error message is printed by anyhow / unravel errors).  
```
Failed to build the site
Error: The anchor in the link `@/post.md#foo` in content/post.md does not exist.
```
2. (Current behavior in this PR) Collect related errors and print them with `console::error`, then bail, passing a "summary" message to anyhow.
```rust
for err in errors { console::error(err) }
return Err(anyhow!("Broken anchor links were found"));
```
Looks like:
```
Failed to build the site
Error: The anchor in the link `@/post.md#foo` in content/post.md does not exist.
Error: The anchor in the link `@/post.md#bar` in content/post.md does not exist.
Error: The anchor in the link `@/post.md#baz` in content/post.md does not exist.
Error: Broken anchor links were found
```
4. (Previous behavior in this PR) Go back to my previous implementation which used `join` judiciously to apply prefixes to 
```rust
return Err(anyhow!(errors.join("\nError: ")));
```
Looks like:
```
Failed to build the site
Error: The anchor in the link `@/post.md#foo` in content/post.md does not exist.
Error: The anchor in the link `@/post.md#bar` in content/post.md does not exist.
Error: The anchor in the link `@/post.md#baz` in content/post.md does not exist.
```
5. Some other solutions I haven't thought of?

Sorry for writing so much, this PR has gone down several different paths and I'm trying to get my bearings.

Reply:
> Sorry for writing so much, this PR has gone down several different paths and I'm trying to get my bearings.

No worries!
I was thinking about it today as well and my thoughts might be a bit out of scope for the PR.

What if the link checker was not using `Error`s but was instead using a `Vec<String>` string for the errors.
This way we can build the error message ourselves nicely, something like:

```
Failed to build the site
Error: Found 3 broken anchor links
  1. The anchor in the link `@/post.md#foo` in content/post.md does not exist.
  2. The anchor in the link `@/post.md#bar` in content/post.md does not exist.
  3. The anchor in the link `@/post.md#baz` in content/post.md does not exist.
```

We can/should obviously apply the same kind of .formatting to external link errors obviously.
What do you think?

Reply:
I can imagine `site/src/link_checking.rs` functions `check_external_links` and `check_internal_links_with_anchors` emitting a `Vec<String>` which contains the text for all the errors or warnings that were encountered.  Those functions are called from `site/src/lib.rs`, so I suppose that's where the logging could happen.  Seems like a clean solution; the logging code would only need to exist in `site/src/lib.rs` instead of being duplicated at every early-return site.  It would potentially change the ordering of error/warning logs relative to regular printlns since the errors/warnings would be delayed until the function returns, but the link checker hardly has any printlns so I don't see a problem there.  Seems good to me, so I guess I'll give it a try next week.

Reply:
The internal link checking that happens in `markdown.rs` would need a similar treatment I suppose, so the code would be more-or-less duplicated, but only in two places. :thinking:

Reply:
Yes the markdown one is unfortunate... maybe keep it as is for now and see if it feels odd when trying it out.

Reply:
It seems pretty good.  The `markdown.rs` behave slightly differently but it doesn't feel odd when using zola.  Here's a demo.

https://asciinema.org/a/Ructp25lMiFB1ztSatUr74ASn